File: Project-2/proj2frontend/WebContent/_user/usercontroller.js
app.controller('UserController',function($scope,$rootScope,$cookieStore,$location,UserService){
	$scope.user={id:'',username:'',password:'',email:'',role:'',isOnline:'',enabled:''};
	$scope.message;
	
	$scope.submit=function(){
		console.log('Entering - submit function in usercontroller')
		UserService.authenticate($scope.user)
		.then(function(response){
				$scope.user=response.data;
				$rootScope.currentUser=$scope.user;
				//key and value pair 
				$cookieStore.put('currentUser',$rootScope.currentUser)
				console.log('currentUser in rootScope ' + $rootScope.currentUser.id)
				$location.path("/home");
		},
		function(response){//invalid user name and password - failure 
			 console.log('invalid username and password')
			  $scope.message="Invalid Username and Password";
			  $location.path("/login");
		})
	}
	
	$scope.registerUser=function(){
		console.log('entering registerUser')
		UserService.registerUser($scope.user)
		.then(function(response){ //success 
			//response.data => user object
			console.log("registration success" + response.status)
			$scope.message="Registration successfull.. login using username and password.."
			$location.path("/login");
		},function(response){//failure
			console.log("registration failed" + response.status)
			//response.data => error object
			console.log(response.data)
			$scope.errorMessage="Registration failed...." + response.data.message
			$location.path("/register")
		})
	}
	
	$scope.friendRequest=function(username){
		alert('friendRequest in userController')
		console.log('friendrequest function')
		UserService.friendRequest(username)
		.then(function(response){
			console.log(response.status);
			alert('Friend request Send')
			getAllUsers();
			$location.path("/getAllUsers")
			
		},
		function(response){
			console.log(response.status);
		}
		)
	}
	
	
	
	function getAllUsers(){
		console.log('entering get all users ')
		UserService.getAllUsers()
		.then(function(response){
		console.log(response.status)
		console.log(response.data)
		$scope.users=response.data
		},function(response){
			console.log(response.status)
		}
		)
	}
	getAllUsers()
	
	
	
})

File: Project-2/proj2backend/src/main/java/com/niit/backend/controllers/UserController.java
package com.niit.backend.controllers;


import java.io.File;
import java.io.FileOutputStream;
import java.util.List;

import javax.servlet.http.HttpSession;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import com.niit.backend.dao.FileUploadDAO;
import com.niit.backend.dao.UserDao;
import com.niit.backend.model.User;
import com.niit.backend.model.Error;
import com.niit.backend.model.UploadFile;
@RestController
public class UserController {
Logger logger=LoggerFactory.getLogger(this.getClass());
@Autowired
private UserDao userDao;
@Autowired
private FileUploadDAO fileUploadDao;

@RequestMapping(value="/login",method=RequestMethod.POST)
public ResponseEntity<?> login(@RequestBody User user,HttpSession session ){
	logger.debug("Entering USERCONTROLLER : LOGIN");
	logger.debug("USERNAME:" + user.getUsername() + " PASSWORD " + user.getPassword() );
	User validUser=userDao.authenticate(user);
	if(validUser==null){
		logger.debug("validUser is null");
		Error error=new Error(1,"Username and password doesnt exists...");
		return new ResponseEntity<Error>(error,HttpStatus.UNAUTHORIZED); //401
	}
	else{
		//key [user] value [validUser- an object of type User]
		//user = [james,123,james@xyz.com,student,true,true]
		session.setAttribute("user", validUser);
		validUser.setOnline(true);
		userDao.updateUser(validUser); // to update online status in db
		logger.debug("validUser is not null");
		
		 //select * from proj2_profile_pics where username='adam';
		  UploadFile getUploadFile=fileUploadDao.getFile(user.getUsername());
		  if(getUploadFile!=null){
	  	String name=getUploadFile.getFileName();
	  	System.out.println(getUploadFile.getData());
	  	byte[] imagefiles=getUploadFile.getData();
	  	try{
	  		String path="C:/eclipsefiles/Project2/proj2backend/src/main/webapp/WEB-INF/resources/images/"+user.getUsername();
	  		File file=new File(path);
	  		//file.mkdirs();
	  		FileOutputStream fos = new FileOutputStream(file);//to Write some data 
	  		fos.write(imagefiles);
	  		fos.close();
	  		}catch(Exception e){
	  		e.printStackTrace();
	  		}
		  }
		
		return new ResponseEntity<User>(validUser,HttpStatus.OK);//200
	}
}

//'?'  - Any Type [User,Error] 
//ENDPOINT : http://localhost:8080/proj2backend/register 
@RequestMapping(value="/register",method=RequestMethod.POST)
public ResponseEntity<?> registerUser(@RequestBody User user){
	//client will send only username, password, email, role 
	try{
	logger.debug("USERCONTROLLER=>REGISTER " + user);
	user.setStatus(true);
	user.setOnline(false);
	User savedUser=userDao.registerUser(user);
	logger.debug("User Id generated is " + savedUser.getId());
	if(savedUser.getId()==0){
		Error error=new Error(2,"Couldnt insert user details ");
		return new ResponseEntity<Error>(error , HttpStatus.CONFLICT);
	}
	else
		return new ResponseEntity<User>(savedUser,HttpStatus.OK);
	}catch(Exception e){
		e.printStackTrace();
		Error error=new Error(2,"Couldnt insert user details. Cannot have null/duplicate values " + e.getMessage());
		return new ResponseEntity<Error>(error , HttpStatus.INTERNAL_SERVER_ERROR);
	}
}
@RequestMapping(value="/logout",method=RequestMethod.PUT)
public ResponseEntity<?> logout(HttpSession session){
	User user=(User)session.getAttribute("user");
	if(user!=null){
		user.setOnline(false);
		userDao.updateUser(user);
		try{
                        //change according to your workspace path and project name
	  		String path="C:/eclipsefiles/Project2/proj2backend/src/main/webapp/WEB-INF/resources/images/"+user.getUsername();
	  		File file=new File(path);
	  		System.out.println(file.delete());
	  		
	  		}catch(Exception e){
	  		e.printStackTrace();
	  		}
	}
	session.removeAttribute("user");
	session.invalidate();
	return new ResponseEntity<Void>(HttpStatus.OK);
}
@RequestMapping(value="/getUsers",method=RequestMethod.GET)
public ResponseEntity<?> getAllUsers(HttpSession session){
	User user=(User)session.getAttribute("user");
	if(user==null)
	return new	ResponseEntity<Error>(new Error(1,"Unauthorized user"),HttpStatus.UNAUTHORIZED);
	else
	{
		List<User> users=userDao.getAllUsers(user);
		for(User u:users)
			System.out.println("IsONline " + u.isOnline());
		return new ResponseEntity<List<User>>(users,HttpStatus.OK);
	}
}

}